### PR TITLE
using system classloader loading class when default classloader cannot find the class

### PR DESCRIPTION
**Pull request by CharlesZ-Chen**
in `AnnotationUtils#getElementValueClass`, add a second try to load class when the default class loader cannot find the class.

This could enable checker-framework loading third party checker's qualifier (e.g. `@Ontology` in Ontology System) as an element value (e.g. `@PolymorphicQualifier` needs the top qualifier in the real type system as it's value)

### CODE REVIEW

**Comment by wmdietl** [comment]
Thanks for looking into this issue!
1. Where can we best document when to use just `Class.forName` and when to also provide `ClassLoader.getSystemClassLoader()`?
2. Why do we need to try both? Wouldn't it be enough to just use the SystemClassLoader?
3. Can you look at all other `Class.forName` uses?

**Comment by CharlesZ-Chen** [comment]
Hi Prof. Dietl,

> 1. Where can we best document when to use just `Class.forName` and when to also provide `ClassLoader.getSystemClassLoader()`?

Generally using `Class.forName` would be enough, and for some special cases (e.g requiring loading class outside from Checker-framework) we might want to also try `SystemClassLoader`.

Maybe we document this at places where might need support loading third party classes (e.g. `AnnotationUtils`)? For other places defaulting is using `Class.forName`.

> 2. Why do we need to try both? Wouldn't it be enough to just use the SystemClassLoader?

Only try `SystemClassLoader` would cause `NoClassFound` exception when using `NullnessChecker`. See travis log here:

https://travis-ci.org/opprop/checker-framework/jobs/157234325#L6546

The reason is because `NullnessChecker` use `URLClassLoader` loads it's qualifiers, and that `URLClassLoader` is a top class loader which doesn't have a parent class loader. Thus we cannot use `SystemClassLoader` loads `NullnessChecker`'s qualifiers.

> 3. Can you look at all other Class.forName uses?

Sure, there are many uses of `Class.forname` in checker-framework, from fundamental classes (e.g. `AnnotationUtils`, `AnnotationClassloader`) to checker-specific classes (e.g. `ValueCheckerUtils`). I still not sure which use is safe and which use might has a potential problem..

File: javacutil/src/org/checkerframework/javacutil/AnnotationUtils.java [hunk 1]
```diff
@@ -576,7 +576,11 @@ public static Name getElementValueClassName(
             AnnotationMirror anno, CharSequence name, boolean useDefaults) {
         Name cn = getElementValueClassName(anno, name, useDefaults);
         try {
-            Class<?> cls = Class.forName(cn.toString());
+            ClassLoader appClassLoader = AnnotationUtils.class.getClassLoader();
```
Comment (wmdietl) [inline javacutil/src/org/checkerframework/javacutil/AnnotationUtils.java]: This is the same logic as contained in `AnnotationClassLoader.getAppClassLoader`.
Can we change that method to take as parameter for which to determine the classloader, make the method static, and invoke that method here?
Then we have one method that we should use in all places where `Class.forName` is invoked.

**Comment by wmdietl** [comment]
This change has been merged in upstream https://github.com/typetools/checker-framework/pull/938
Closing this PR.

**Pull request closed without merging**